parler_tts/modeling_parler_tts.py

```diff
@@ -829,23 +1320,55 @@ def forward(
         else:
             raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
 
-        # past_key_values_length
-        past_key_values_length = past_key_values[0][0].shape[2] if past_key_values is not None else 0
-
         if inputs_embeds is None:
             inputs_embeds = sum([self.embed_tokens[codebook](input[:, codebook]) for codebook in range(num_codebooks)])
 
         # if prompt_hidden_states, fuse to inputs_embeds and update input shape
         if prompt_hidden_states is not None:
             inputs_embeds = torch.cat([prompt_hidden_states, inputs_embeds], dim=1)
+            # this step is done at first generation step
+            # cache_position is initialized in _get_initial_cache_position (GenerationMixin method),
```

Comment:
Indeed, it would be cleaner! I surmised that since `_sample` in [GenerationMixin](https://github.com/huggingface/transformers/blob/85a1269e19af022e04bc2aad82572cd5a9e8cdd9/src/transformers/generation/utils.py#L357) calls `_get_initial_cache_position`, the only way to overwrite `_get_initial_cache_position` would have been to copy the entire `_sample` method into our modeling file to have it call our `_get_initial_cache_position`. However, I was mistaken about the method resolution order in the case of a Mixin design pattern. Thanks a lot for this idea and for teaching me this! I'll implement it this way.

Reply:
Since we are building `inputs_embeds` [here](https://github.com/eustlb/parler-tts/blob/ad25e2ba355ea47a42e85feff77e8dab8039ae1a/parler_tts/modeling_parler_tts.py#L2895C45-L2895C58), `inputs_embeds` has the good shape before passing to `_get_initial_cache_position` and the `cache_position` is initialized as it is supposed to be [here](https://github.com/huggingface/transformers/blob/df6eee9201e4ba2b80cea021a18e95ada26ca2cc/src/transformers/generation/utils.py#L1425), so the proposed change is actually not necessary (no need to overwrite `_get_initial_cache_position`). 

Nevertheless, I had to hard code `cache_position` as I did because since the `prompt_hidden_state` is prepended to `inputs_embeds` during the forward, the number of `decoder_input_ids` passed to `prepare_inputs_for_generation` [here](https://github.com/huggingface/transformers/blob/df6eee9201e4ba2b80cea021a18e95ada26ca2cc/src/transformers/generation/utils.py#L2925) is not the good one and thus cache_position will not be what it is supposed to be (because of [this line](https://github.com/eustlb/parler-tts/blob/ad25e2ba355ea47a42e85feff77e8dab8039ae1a/parler_tts/modeling_parler_tts.py#L2800))

That is why cache_position has either to be hardcoded in the forward has I did, either directly adapt `prepare_inputs_for_generation`. I implement the latter in [this commit](https://github.com/huggingface/parler-tts/pull/89/commits/f29392bccafece9630eec14c2e0b6da02a990fe5) as I agree that it is more elegant than hard coding it in the forward pass. Thank you!